core/templates/pages/exploration-editor-page/services/exploration-data.service.ts

```diff
@@ -17,186 +17,212 @@
+import { tap } from 'rxjs/operators';
+import { AlertsService } from 'services/alerts.service';
+import { LoggerService } from 'services/contextual/logger.service';
+import { UrlService } from 'services/contextual/url.service';
+import { WindowRef } from 'services/contextual/window-ref.service';
+import { LocalStorageService } from 'services/local-storage.service';
+import { ExplorationDataBackendApiService } from './exploration-data-backend-api.service';
 
-require('services/services.constants.ajs.ts');
+export interface DraftAutoSaveResponse {
+  'draft_change_list_id': number;
+  'is_version_of_draft_valid': boolean;
+}
+@Injectable({
+  providedIn: 'root'
+})
+export class ExplorationDataService {
+  draftChangeListId: number | null = null;
+  explorationDraftAutosaveUrl: string;
+  explorationId: string;
+  resolvedAnswersUrlPrefix: string;
+  data: ExplorationBackendDict;
 
-angular.module('oppia').factory('ExplorationDataService', [
-  '$http', '$q', 'AlertsService',
-  'EditableExplorationBackendApiService', 'LocalStorageService',
-  'LoggerService', 'ReadOnlyExplorationBackendApiService',
-  'UrlService', 'WindowRef',
-  function(
-      $http, $q, AlertsService,
-      EditableExplorationBackendApiService, LocalStorageService,
-      LoggerService, ReadOnlyExplorationBackendApiService,
-      UrlService, WindowRef) {
-    // The pathname (without the hash) should be: .../create/{exploration_id}
-    var explorationId = '';
-    var draftChangeListId = null;
-    var pathname = UrlService.getPathname();
-    var pathnameArray = pathname.split('/');
-    for (var i = 0; i < pathnameArray.length; i++) {
+  constructor(
+    private alertsService: AlertsService,
+    private editableExplorationBackendApiService:
+      EditableExplorationBackendApiService,
+    private explorationDataBackendApiService: ExplorationDataBackendApiService,
+    private localStorageService:LocalStorageService,
+    private loggerService: LoggerService,
+    private readOnlyExplorationBackendApiService:
+      ReadOnlyExplorationBackendApiService,
+    private urlService: UrlService,
+    private windowRef: WindowRef
+  ) {
+    let explorationId = '';
+    // The pathname (without the hash) should be: .../create/{exploration_id}.
+    const pathname = this.urlService.getPathname();
+    const pathnameArray = pathname.split('/');
+    for (let i = 0; i < pathnameArray.length; i++) {
       if (pathnameArray[i] === 'create') {
         explorationId = pathnameArray[i + 1];
         break;
       }
     }
 
     if (!explorationId) {
-      LoggerService.error(
-        'Unexpected call to ExplorationDataService for pathname ',
-        pathname);
-      // Note: if we do not return anything, Karma unit tests fail.
-      return {};
+      this.loggerService.error(
+        'Unexpected call to ExplorationDataService for pathname: ' + pathname);
+      this.autosaveChangeList = undefined;
```

Comment:
Hi @srijanreddy98 -- I suspect that these lines might be the cause of a bunch of common flakes we are getting in the frontend tests along the lines of "TypeError: this.explorationDataService.autosaveChangeListAsync is not a function".

Could you please provide some insight on why they were added? Thanks!

/cc @U8NWXD

Reply:
https://github.com/oppia/oppia/pull/11564/files#r706660354

Reply:
Ah, I see -- the returned dict is the list of functions? Got it, thanks.